src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -117,14 +117,31 @@ private static bool IsFixableNode(SyntaxNode node)
                 };
             }
 
-            if (TryGetCommonReceiver(leftReceiver, rightReceiver, model) is var (commonReceiver, leftNames, rightNames))
+            if (TryGetCommonReceiver(leftReceiver, rightReceiver, leftTarget, rightTarget, model) is var (commonReceiver, leftNames, rightNames))
             {
                 return root =>
                 {
-                    var leftSubpattern = CreateSubpattern(leftNames, CreatePattern(leftReceiver, leftTarget, leftFlipped));
-                    var rightSubpattern = CreateSubpattern(rightNames, CreatePattern(rightReceiver, rightTarget, rightFlipped));
-                    var replacement = IsPatternExpression(commonReceiver, RecursivePattern(leftSubpattern, rightSubpattern));
-                    return root.ReplaceNode(logicalAnd, AdjustBinaryExpressionOperands(logicalAnd, replacement));
+                    // It's possible we decided to discard a pattern due to it being redundant (such as a null check
+                    // combined with a property check belonging to the same field we confirmed not being null)
+                    if (leftNames.Length == 0)
+                    {
+                        var rightSubpattern = CreateSubpattern(rightNames, CreatePattern(rightReceiver, rightTarget, rightFlipped));
+                        var replacement = IsPatternExpression(commonReceiver, RecursivePattern(rightSubpattern));
+                        return root.ReplaceNode(logicalAnd, AdjustBinaryExpressionOperands(logicalAnd, replacement));
+                    }
+                    else if (rightNames.Length == 0)
+                    {
+                        var leftSubpattern = CreateSubpattern(leftNames, CreatePattern(leftReceiver, leftTarget, leftFlipped));
+                        var replacement = IsPatternExpression(commonReceiver, RecursivePattern(leftSubpattern));
+                        return root.ReplaceNode(logicalAnd, AdjustBinaryExpressionOperands(logicalAnd, replacement));
+                    }
```

Comment:
i'm on the edge of wanting a local helper here, since tehse if/else blocks are nearly identical.  can you try taht out?  if it doesn't help, can keep like this.

Reply:
I tried: it seemed to make it a bit messy (it involves a lot of parameters) and the case where both left and right are used makes it hard to extract so I will leave as is.